---
title: "Role of Piwi-piRNA pathway in somatic and cancer cells"
author: "__Konstantinos Geles__"
date: "Thu Jun 30 2022, Last Update: `r format(Sys.Date(), '%a %b %d %Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 3
    df_print: paged
  pdf_document:
    toc: yes
    toc_depth: 3
  html_notebook: null
editor_options:
  chunk_output_type: console
subtitle: UMG PhD Programme of Molecular and Translational Oncology - Circle XXXIV
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

This project contains the scripting part of the Doctoral Dissertation of **Konstantinos Geles** with doi:   

# CHAPTER 1: Role of the PIWI-piRNA pathway in Colorectal Cancer (CRC)  
  
## 1.3 Evaluation of piRNA expression in CRC cell lines and comparison to germline 

Import Libraries
```{r}
library(readxl)
library(dplyr)
library(tibble)
library(tidyr)
library(ComplexHeatmap)
library(circlize)
library(purrr)
library(stringr)
library(ggplot2)
library(scales)
```

## sncRNA concetrations in CRC cell lines and testis samples
Using the data on the url: we import the expression values for each cell line.
Import the table
```{r}
sncRNA_CRC_cells <- readxl::read_xlsx("Chapter_1_3/Table_S3_Sellitto_et_al.xlsx",
                                     skip = 4,n_max = 10) %>% 
    select(-c(1,3:4)) %>% 
    mutate(Annotation = str_replace(Annotation, "Human_Testis_Pool", "Testis"))
```

# custom theme for the plots
```{r}
# Use colourblind-friendly colours
friendly_cols <- viridisLite::plasma(10)

# Set theme
custom_theme <-
  list(
    #scale_fill_manual(values = friendly_cols),
    #scale_color_manual(values = friendly_cols),
    scale_fill_brewer(palette = "Set1"),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 20),
        #legend.position = "none",
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        plot.title = element_text(hjust = 0.5),
      )
  )
```

create the plot
```{r}
p_sncRNA <- sncRNA_CRC_cells %>% 
    rename("Cell line"= Annotation) %>% 
    pivot_longer(cols = -"Cell line", names_to = "Database", values_to = "Reads") %>% 
    ggplot() +
    geom_col(mapping = aes(x = `Cell line`, y = Reads, fill = Database),position = "fill") +
    ylab("Mapped Reads")+
    ggtitle("Percentage of mapped reads to each sncRNA database \nfor the CRC cell-lines and testis samples")+
    scale_y_continuous(labels = scales::percent)+
    custom_theme

tiff(filename = file.path("FIG_15_sncRNA_DBs_CRC_lines_Testis.tiff"),
     compression = "none", height = 10, width = 14,  units = 'in', res = 600)
p_sncRNA
dev.off()
```

###  piRNA expression in CRC cell lines and testis samples
For this analysis we have use SPORTS 1.0 to perform all the preprocessing, alignment and quantification steps as shown in the ... GitHub Repository.
Using TMM normalization and transforming to CPM we arrive to a table that is published in Sellitto et al as supplementary material.
At this stage we will create a Heatmap showing the expression values between samples.

Import the table
```{r}
piRNA_CRC_cells <- readxl::read_xlsx("Chapter_1_3/Table_S4_Sellitto_et_al.xlsx",
                                     skip = 4)
```

find the most variant piRNAs and the the common expressed in all samples
```{r}
MVariant_piRNA <- piRNA_CRC_cells %>% 
    rowwise() %>%
    mutate(mad_piRNA = mad(c_across(-smallRNA))) %>% 
    arrange(desc(mad_piRNA)) 

common_high_piRNA <- piRNA_CRC_cells %>% 
    pivot_longer(cols = -smallRNA) %>% 
    filter(value > 1 ) %>% 
    count(smallRNA) %>% 
    filter(n > 8 ) %>% 
    pull(smallRNA)

common_piRNA <- piRNA_CRC_cells %>%
    filter(smallRNA %in% common_high_piRNA)
```

keep the 500 most variant piRNA
```{r}
MVariant_piRNA_500 <- MVariant_piRNA %>%
    select(-mad_piRNA) %>% 
    head(500)
```

reshape the tables to matrices
```{r}
mat_v <- MVariant_piRNA_500 %>%
    column_to_rownames("smallRNA") %>% 
    as.matrix()

mat_com <- common_piRNA %>%
    filter(!smallRNA %in% MVariant_piRNA_500$smallRNA) %>% 
    column_to_rownames("smallRNA") %>% 
    as.matrix()
```

Scale the matrix for the heatmap
```{r}
# make the matrices for the heatmap -----
sc_mat <- mat_v %>% t() %>% scale() %>% t()
sc_mat %>% dim()
sc_mat %>% head()
hist(sc_mat)


sc_mat_c <- mat_com %>% t() %>% scale() %>% t()
sc_mat_c %>% dim()
sc_mat_c %>% head()
hist(sc_mat_c)
```

make the heatmap
```{r}
f <- colorRamp2(c(-1, 0 ,1),c("#53868B", "#8B8878", "#FFD700"))

ha_1 <- HeatmapAnnotation(Cell_lines = colnames(sc_mat),
            col = list(Cell_lines =
                           c(wesanderson::wes_palettes$Rushmore[1:4],
                             wesanderson::wes_palettes$Darjeeling1) %>% 
                           set_names(colnames(sc_mat))))

ht_1 <- Heatmap(matrix = sc_mat, #data
        top_annotation = ha_1, #annot
        col = f, #colors data  
        row_title_gp = gpar(fontsize = 20),
        column_title_gp = gpar(fontsize = 20),
        column_names_gp = gpar(fontsize = 20),
        column_names_rot = 45,
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 15, fontface = "bold"),
            labels_gp = gpar(fontsize = 15)),
        show_row_dend = TRUE,
        show_row_names = FALSE,
        show_column_names = TRUE,
        name = "z-score \nexpression",
        clustering_distance_columns = "spearman",
        clustering_method_columns =  "ward.D2",
        clustering_method_rows = "ward.D2",
        clustering_distance_rows = "spearman",
        row_dend_reorder = TRUE,
        row_title = "piRNAs",
        column_title = str_glue("Heatmap of {nrow(sc_mat)} most variant piRNAs\nbetween cell lines"),
        
)

ht_c <- Heatmap(matrix = sc_mat_c, #data
        top_annotation = ha_1, #annot
        col = f, #colors data    
        row_title_gp = gpar(fontsize = 20),
        column_title_gp = gpar(fontsize = 20),
        column_names_gp = gpar(fontsize = 20),
        column_names_rot = 45,
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 15, fontface = "bold"),
            labels_gp = gpar(fontsize = 15)),
        show_row_dend = TRUE,
        show_row_names = FALSE,
        show_column_names = TRUE,
        name = "z-score \nexpression",
        clustering_distance_columns = "spearman",
        clustering_method_columns =  "ward.D2",
        clustering_method_rows = "ward.D2",
        clustering_distance_rows = "spearman",
        row_dend_reorder = TRUE,
        column_title = str_glue("Heatmap of {nrow(sc_mat_c)} common, most expressed piRNAs\nbetween cell lines")
)

tiff(filename = file.path("FIG_16_piRNA_most_var_CRC_lines_Testis.tiff"),
     compression = "none", height = 10, width = 14,  units = 'in', res = 600)
ht_1
dev.off()

tiff(filename = file.path("FIG_16_piRNA_most_expr_CRC_lines_Testis.tiff"),
     compression = "none", height = 10, width = 14,  units = 'in', res = 600)
ht_c
dev.off()
```



