---
title: "Role of Piwi-piRNA pathway in somatic and cancer cells"
author: "__Konstantinos Geles__"
date: "Wed Jul 13 2022, Last Update: `r format(Sys.Date(), '%a %b %d %Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 3
    df_print: paged
  pdf_document:
    toc: yes
    toc_depth: 3
  html_notebook: null
editor_options:
  chunk_output_type: console
subtitle: UMG PhD Programme of Molecular and Translational Oncology - Circle XXXIV
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

This project contains the scripting part of the Doctoral Dissertation of **Konstantinos Geles** with doi:   

# CHAPTER 3: Data Analysis Workflow for small-RNAseq focused on piRNAs  
  
## 3.1.2 Analysis of samples from patients with CRC and comparison with previous analyses.

###  DE sncRNAs 

Import Libraries
```{r}
library(vroom)
library(dplyr)
library(tidyr)
library(purrr)
library(tibble)
library(stringr)
library(ggplot2)
library(scales)
library(ggpmisc)
library(readr)
library(edgeR)
library(ComplexHeatmap)
library(circlize)

library(clusterProfiler)
library(wppi)
library(ReactomePA)
library(enrichplot)
library(forcats)
```

import the resulted de sncRNAs of CRC patients
```{r}
Cardio_SC <- vroom("Chapter_3/DEA_piRNA_GSE161081_cardioMouse_GRCh38_21_Jul_2022/GSE161081_cardioMouse_all_comparisons_long_voom_TMMQW_salmon_fc_LFCs_21_Jul_2022.txt") %>% 
    mutate(contrast = case_when(
        contrast == "iCM_v_CM" ~ "iCM v. aCM",
        contrast == "iCM_v_CSC" ~ "iCM v. CSC",
        contrast == "CM_v_CSC" ~ "aCM v. CSC",
    ))
```

PhD theme for plots
```{r}
wes_cols <- c(wesanderson::wes_palettes$Cavalcanti1,
              wesanderson::wes_palettes$Royal2,
              wesanderson::wes_palettes$Darjeeling2)

PhD_theme <-
  list(
    scale_fill_manual(values = wes_cols),
    #scale_color_manual(values = wes_cols),
    #scale_fill_brewer(palette = "Set2"),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(size = 20, colour = "black"),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 20),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10),
                                    colour = "black"),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10),
                                    colour = "black"),
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 20),
        axis.text.y = element_text(size = 20, colour = "black"),
        plot.title = element_text(hjust = 0.5, colour = "black"),
        legend.position = "bottom", 
        legend.direction = "horizontal" 
      )
  )
```

plot the amount of DE per sncRNA class
```{r}
plot_Cardio_SC_DE <- Cardio_SC %>% 
    filter(adj.P.Val < 0.05, !is.na(gene_type), contrast != "aCM v. CSC") %>% 
    ggplot(aes(x = gene_type, fill = gene_type)) +
    geom_bar(stat = "count") +
    stat_count(geom = "text", colour = "black", size = 7.5, 
              aes(label = ..count..), position  = position_fill(vjust = 700 )) +
    facet_wrap(facets = vars(contrast, quantification), ncol = 2) + 
    xlab("sncRNA Class") + 
    ggtitle("Amount of sncRNAs found DE between quantification approaches\n in mouse CardioMyocytes and Cardiac Stem Cells samples") +
    PhD_theme

tiff(filename = file.path("FIG_45_cardio_mouse_DE_salmon_FC.tiff"),
     compression = "none", height = 16, width = 16,  units = 'in', res = 600)
plot_Cardio_SC_DE
dev.off()
```

###  DE piRNAs

DE piRNAs, keep the union of both approaches
```{r}
DE_piRNA <- Cardio_SC %>% 
    filter(adj.P.Val < 0.05, gene_type == "piRNA"| is.na(gene_type)) %>% 
    select(contrast, smallRNA:logFC) %>% 
    pivot_wider(names_from = c(contrast, quantification), values_from = logFC) 

DE_piRNA %>% drop_na()

iCM_v_CM_piRNA <- DE_piRNA %>% 
    select(smallRNA, starts_with("iCM v. a")) %>% 
    drop_na() %>% 
    filter(!(if_any(c(-smallRNA), ~ .x > 0) & if_any(c(-smallRNA), ~ .x < -0.1)|
               if_any(c(-smallRNA), ~ .x < -0.1) & if_any(c(-smallRNA), ~ .x > 0))) 
    
iCM_v_CSC_piRNA <- DE_piRNA %>% 
    select(smallRNA,  starts_with("iCM v. CSC")) %>% 
    drop_na() %>% 
    filter(!(if_any(c(-smallRNA), ~ .x > 0) & if_any(c(-smallRNA), ~ .x < -0.1)|
               if_any(c(-smallRNA), ~ .x < -0.1) & if_any(c(-smallRNA), ~ .x > 0)))
```

import the dgl for the downstream visualization of expression values
```{r}
dgl_fc_TMM <- read_rds("Chapter_3/EDA_mouse_cardiomyocytes_GRCm38_21_Jul_2022/featureCounts/list_norm_dgls_featureCounts.rds") %>% 
    magrittr::extract2("TMM") 

# get the matrix lcpm values

lcpm_mat <- dgl_fc_TMM%>% #get only the TMM normalized expressions
    edgeR::cpm(log = TRUE, prior.count = 4) # transform to lcpm

# subset to DE piRNAs

iCM_v_CM_piRNA_mat <- lcpm_mat[iCM_v_CM_piRNA$smallRNA,]
iCM_v_CSC_piRNA_mat <- lcpm_mat[iCM_v_CSC_piRNA$smallRNA,]
```

import the groups table
```{r}
table_groups <-  dgl_fc_TMM %>% 
    magrittr::extract2("colours")

table_groups <- table_groups %>% mutate(
    group = case_when(
        group == "CM" ~ "aCM",
        group == "CSC_CM" ~ "iCM",
        TRUE ~ as.character(group)
    ) %>% as_factor,
    group_col = case_when(
        group_col == "#4daf4a" ~ "#D8B70A",
        group_col == "#f781bf" ~ "#02401B",
        TRUE ~ as.character(group_col)
    ) %>% as_factor
)
```

#### iCM vs aCM
Scale the matrix for the heatmap
```{r}
# make the exprs matrix for the heatmap -----
iCM_v_CM_mat <- iCM_v_CM_piRNA_mat %>% t() %>% scale() %>% t()

iCM_v_CM_mat %>% dim()
iCM_v_CM_mat %>% head()
str_c("min = ", round(min(iCM_v_CM_mat),3),
      "| median = ", round(median(iCM_v_CM_mat),3), 
      "| max = ", round(max(iCM_v_CM_mat),3))

stopifnot(identical(as.character(table_groups$sample_name), colnames(iCM_v_CM_mat)))

# make the logFC matrix for the heatmap -----
iCM_v_CM_fc_mat <- iCM_v_CM_piRNA %>% column_to_rownames("smallRNA") %>% as.matrix()
iCM_v_CM_fc_mat %>% dim()
iCM_v_CM_fc_mat %>% head()
str_c("min = ", round(min(iCM_v_CM_fc_mat, na.rm = T),3),
      "| median = ", round(median(iCM_v_CM_fc_mat, na.rm = T),3), 
      "| max = ", round(max(iCM_v_CM_fc_mat, na.rm = T),3))

iCM_v_CM_fc_mat <- iCM_v_CM_fc_mat[rownames(iCM_v_CM_mat),]

rownames(iCM_v_CM_mat) <- rownames(iCM_v_CM_mat) %>% str_remove("_GR.+")

rownames(iCM_v_CM_fc_mat) <- rownames(iCM_v_CM_fc_mat) %>% str_remove("_GR.+")

identical(rownames(iCM_v_CM_mat), rownames(iCM_v_CM_fc_mat))
```

make the heatmap
```{r}
f <- colorRamp2(c(-1, 0, 1), c("#53868B", "#8B8878", "#FFD700"))

f_LFCs <- colorRamp2(c(-4, 0, 4),
                    c("#008B8B", "black", "#B22222"))
# mat
ha_1 <- HeatmapAnnotation(
                          Group = table_groups$group,
                          Batch = table_groups$batch,
            col = list(
                Group = table_groups %>%  
                           select(group, group_col) %>% 
                           deframe(),
                Batch = table_groups %>%  
                           select(batch, batch_col) %>% 
                           deframe()),
                annotation_name_side = "left",
            annotation_legend_param = list(
                Group = list(title_gp = gpar( fontsize = 20, fontface = "bold"), 
                       labels_gp = gpar(fontsize = 20)),
                
                Batch = list(title_gp = gpar(fontsize = 20, fontface = "bold"), 
                       labels_gp = gpar(fontsize = 20)))
            )
            

# lFCS
ha_1_LFCs <- HeatmapAnnotation(Method = c("salmon", "FeatureCounts"),
                              col = list(Method = wesanderson::wes_palettes$Moonrise1[c(1,4)] %>%
                          set_names("salmon", "FeatureCounts")), 
                          annotation_legend_param = list(
                Method = list(title_gp = gpar( fontsize = 20, fontface = "bold"), 
                       labels_gp = gpar(fontsize = 20)) )
                )

ht_1 <- Heatmap(
        matrix = iCM_v_CM_mat,          # data
        top_annotation = ha_1,          # annot
        col = f,                        # colors data
        name = "z-score \nabundance",
        
        show_row_dend = TRUE,
        show_column_names = FALSE,
        
        row_names_side = "left",
        row_dend_side = "right",
        
        clustering_method_columns =  "ward.D2",
        clustering_method_rows = "ward.D2",
        clustering_distance_columns = "euclidean",
        clustering_distance_rows = "euclidean",
        
        column_km = 3,
        row_km = 3,
        
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 20, fontface = "bold"),
            labels_gp = gpar(fontsize = 20)),
        
        row_names_gp = grid::gpar(fontsize = 15),
        row_title = "piRNAs",
        row_title_gp = gpar(fontsize = 20, fontface = "bold"),
        
        column_title_gp = gpar(fontsize = 20, fontface = "bold"),
        column_title = str_glue("Heatmap of {nrow(iCM_v_CM_mat)} common piRNAs\nidentified DE in transcriptomic and genomic approaches\nfor the iCM v. aCM comparison")) 

ht_1_lFCs <- Heatmap(matrix = iCM_v_CM_fc_mat, #data
        top_annotation = ha_1_LFCs, #annot
        col = f_LFCs, #colors data   
        
        show_row_dend = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        show_column_dend = FALSE,
        row_dend_reorder = TRUE,
        
        row_dend_side = "right",
        
        name = "Log Fold Change\niCM v. aCM",
        row_names_gp = grid::gpar(fontsize = 15),
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 20, fontface = "bold"),
            labels_gp = gpar(fontsize = 20))
        )

draw(ht_1 + ht_1_lFCs,
      column_title = str_glue("Heatmap of {nrow(iCM_v_CM_mat)} DE piRNAs"), merge_legend = TRUE)

tiff(filename = file.path("FIG_46_piRNA_iCM_v_aCM_common_DE_cardiomyo_patients.tiff"),
     compression = "none", height = 16, width = 16,  units = 'in', res = 600)
draw(ht_1 + ht_1_lFCs , merge_legend = TRUE)
dev.off()

```

#### iCM vs CSC
Scale the matrix for the heatmap
```{r}
# make the exprs matrix for the heatmap -----
iCM_v_CSC_mat <- iCM_v_CSC_piRNA_mat %>% t() %>% scale() %>% t()

iCM_v_CSC_mat %>% dim()
iCM_v_CSC_mat %>% head()

str_c("min = ", round(min(iCM_v_CSC_mat),3),
      "| median = ", round(median(iCM_v_CSC_mat),3), 
      "| max = ", round(max(iCM_v_CSC_mat),3))

stopifnot(identical(as.character(table_groups$sample_name), colnames(iCM_v_CSC_mat)))

# make the logFC matrix for the heatmap -----
iCM_v_CSC_fc_mat <- iCM_v_CSC_piRNA %>% column_to_rownames("smallRNA") %>% as.matrix()
iCM_v_CSC_fc_mat %>% dim()
iCM_v_CSC_fc_mat %>% head()

str_c("min = ", round(min(iCM_v_CSC_fc_mat, na.rm = T),3),
      "| median = ", round(median(iCM_v_CSC_fc_mat, na.rm = T),3), 
      "| max = ", round(max(iCM_v_CSC_fc_mat, na.rm = T),3))

iCM_v_CSC_fc_mat <- iCM_v_CSC_fc_mat[rownames(iCM_v_CSC_mat),]

rownames(iCM_v_CSC_fc_mat) <- rownames(iCM_v_CSC_fc_mat) %>% str_remove("_GR.+")

rownames(iCM_v_CSC_mat) <- rownames(iCM_v_CSC_mat) %>% str_remove("_GR.+")

identical(rownames(iCM_v_CSC_fc_mat), rownames(iCM_v_CSC_mat))
```

make the heatmap
```{r}
f <- colorRamp2(c(-1, 0, 1), c("#53868B", "#8B8878", "#FFD700"))

f_LFCs <- colorRamp2(c(-4, 0, 4),
                    c("#008B8B", "black", "#B22222"))
# mat
ha_1 <- HeatmapAnnotation(
                          Group = table_groups$group,
                          Batch = table_groups$batch,
            col = list(
                Group = table_groups %>%  
                           select(group, group_col) %>% 
                           deframe(),
                Batch = table_groups %>%  
                           select(batch, batch_col) %>% 
                           deframe()),
                annotation_name_side = "left",
            annotation_legend_param = list(
                Group = list(title_gp = gpar( fontsize = 20, fontface = "bold"), 
                       labels_gp = gpar(fontsize = 20)),
                
                Batch = list(title_gp = gpar(fontsize = 20, fontface = "bold"), 
                       labels_gp = gpar(fontsize = 20)))
            )
            

# lFCS
ha_1_LFCs <- HeatmapAnnotation(Method = c("salmon", "FeatureCounts"),
                              col = list(Method = wesanderson::wes_palettes$Moonrise1[c(1,4)] %>%
                          set_names("salmon", "FeatureCounts")), 
                          annotation_legend_param = list(
                Method = list(title_gp = gpar( fontsize = 20, fontface = "bold"), 
                       labels_gp = gpar(fontsize = 20)) )
                )

ht_1 <- Heatmap(
        matrix = iCM_v_CSC_mat,          # data
        top_annotation = ha_1,          # annot
        col = f,                        # colors data
        name = "z-score \nabundance",
        
        show_row_dend = TRUE,
        show_column_names = FALSE,
        
        row_names_side = "left",
        row_dend_side = "right",
        
        clustering_method_columns =  "ward.D2",
        clustering_method_rows = "ward.D2",
        clustering_distance_columns = "euclidean",
        clustering_distance_rows = "euclidean",
        
        column_km = 3,
        row_km = 3,
        
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 20, fontface = "bold"),
            labels_gp = gpar(fontsize = 20)),
        
        row_names_gp = grid::gpar(fontsize = 15),
        row_title = "piRNAs",
        row_title_gp = gpar(fontsize = 20, fontface = "bold"),
        
        column_title_gp = gpar(fontsize = 20, fontface = "bold"),
        column_title = str_glue("Heatmap of {nrow(iCM_v_CSC_mat)} common piRNAs\nidentified DE in transcriptomic and genomic approaches\nfor the iCM v. CSC comparison")) 

ht_1_lFCs <- Heatmap(matrix = iCM_v_CSC_fc_mat, #data
        top_annotation = ha_1_LFCs, #annot
        col = f_LFCs, #colors data   
        
        show_row_dend = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        show_column_dend = FALSE,
        row_dend_reorder = TRUE,
        
        row_dend_side = "right",
        
        name = "Log Fold Change\niCM v. CSC",
        row_names_gp = grid::gpar(fontsize = 15),
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 20, fontface = "bold"),
            labels_gp = gpar(fontsize = 20))
        )

draw(ht_1 + ht_1_lFCs,
      column_title = str_glue("Heatmap of {nrow(iCM_v_CSC_fc_mat)} DE piRNAs"), merge_legend = TRUE)

tiff(filename = file.path("FIG_47_piRNA_common_iCM_v_CSC_DE_cardiomyo_patients.tiff"),
     compression = "none", height = 16, width = 16,  units = 'in', res = 600)
draw(ht_1 + ht_1_lFCs , merge_legend = TRUE)
dev.off()

```

### Commmon piRNAs signif DE

make a heatmap of logfcs
```{r}
# make the logFC matrix for the heatmap -----
all_lFCs <- iCM_v_CM_piRNA %>% 
    inner_join(iCM_v_CSC_piRNA) %>% 
    column_to_rownames("smallRNA") %>% 
    as.matrix()
all_lFCs %>% dim()
all_lFCs %>% head()

str_c("min = ", round(min(all_lFCs, na.rm = T),3),
      "| median = ", round(median(all_lFCs, na.rm = T),3), 
      "| max = ", round(max(all_lFCs, na.rm = T),3))

rownames(all_lFCs) <- rownames(all_lFCs) %>% str_remove("_GR.+")

f_all_LFCs <- colorRamp2(c(-2, 0, 2),
                    c("#008B8B", "black", "#B22222"))

# lFCS
comparisons <- colnames(all_lFCs) %>% str_remove("_.+")
methodsss <- colnames(all_lFCs) %>% str_remove(".+_")

ha_all_LFCs <- HeatmapAnnotation(Comparison = comparisons,
                               Method = methodsss,
                              col = list(Comparison = wesanderson::wes_palettes$Zissou1[c(1,4)] %>%
                                             rep(each = 2) %>% 
                                             set_names(comparisons),
                                         Method = wesanderson::wes_palettes$Moonrise1[c(1,4)] %>%
                                             rep(2) %>% 
                                             set_names(methodsss)
                                         ), 
                          annotation_legend_param = list(
                
                              Comparison = list(
                                  title_gp = gpar( fontsize = 20, fontface = "bold"), 
                                 labels_gp = gpar(fontsize = 20)),
                              Method = list(
                                  title_gp = gpar( fontsize = 20, fontface = "bold"), 
                                 labels_gp = gpar(fontsize = 20)))
                )

ht_all_lFCs <- Heatmap(matrix = all_lFCs, #data
        top_annotation = ha_all_LFCs, #annot
        col = f_all_LFCs, #colors data   
        
        show_row_dend = TRUE,
        show_row_names = TRUE,
        show_column_names = FALSE,
        show_column_dend = TRUE,
        row_dend_reorder = TRUE,
        
        row_title = "piRNAs",
        row_dend_side = "left",
        
        name = "Log Fold Change",
        row_names_gp = grid::gpar(fontsize = 15),
        heatmap_legend_param = list(
            title_gp = gpar(fontsize = 20, fontface = "bold"),
            labels_gp = gpar(fontsize = 20)),
        
        column_title_gp = gpar(fontsize = 20, fontface = "bold"),
        column_title = str_glue("Heatmap of {nrow(all_lFCs)} common piRNAs\nidentified DE both comparisons")
        )

tiff(filename = file.path("FIG_48_piRNA_common_all_lFCs_DE_cardiomyo_patients.tiff"),
     compression = "none", height = 16, width = 16,  units = 'in', res = 600)
draw(ht_all_lFCs)
dev.off()
```

###  Gene predicted targets for the DE piRNA from mouse cardiomyocytes

import the table with predicted targets
```{r}
total_pred_targ <- vroom("Chapter_3/piRNA_predicted_Targets.v_GRCm38_BIOC.txt")
```

search the union of DE piRNAs from both FC and salmon
```{r}
total_pred_targ %>% 
    filter(piRNA_id %in% iCM_v_CM_piRNA$smallRNA) %>% 
    distinct(piRNA_id)

piRNA_DE_targets_iCM_v_CM <- total_pred_targ %>% 
    filter(piRNA_id %in% iCM_v_CM_piRNA$smallRNA) %>% 
    distinct(piRNA_id, Target_gene_name, .keep_all = TRUE) 

piRNA_DE_targets_iCM_v_CSC <- total_pred_targ %>% 
    filter(piRNA_id %in% iCM_v_CSC_piRNA$smallRNA) %>% 
    distinct(piRNA_id, Target_gene_name, .keep_all = TRUE)

Entr_Gene_Ids <- bitr(piRNA_DE_targets_iCM_v_CM$Target_gene_name, 
                     fromType = "SYMBOL", 
                     toType = "ENTREZID", OrgDb = "org.Mm.eg.db")

```

```{r}
goenr <- enrichGO(Entr_Gene_Ids$ENTREZID, ont = "MF", OrgDb = 'org.Mm.eg.db', 
                  readable = TRUE) %>%
    as_tibble()

reactome_path <- enrichPathway(gene = Entr_Gene_Ids$ENTREZID, 
                               organism = "mouse",
                               pvalueCutoff = 0.05, readable = TRUE) %>%
    as_tibble()


```


