---
title: "Role of Piwi-piRNA pathway in somatic and cancer cells"
author: "__Konstantinos Geles__"
date: "Mon Jun 20 2022, Last Update: `r format(Sys.Date(), '%a %b %d %Y')`"
output:
  html_notebook:
  pdf_document:
    toc: yes
    toc_depth: 3
editor_options:
  chunk_output_type: console
subtitle: "UMG PhD Programme of Molecular and Translational Oncology - Circle XXXIV"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

-   This project contains the scripting part of the Doctoral Dissertation of **Konstantinos Geles** with doi: ## CHAPTER 1: Role of the PIWI-piRNA pathway in Colorectal Cancer (CRC) ## 1.1 Integrative data analysis of gene expression and DNA methylation data from public datasets

To investigate the functional role of the PIWI-piRNA pathway, I queried various public genomic data repositories.
First I looked into the Genotype-Tissue Expression (GTEx) project and downloaded the data of GTEx Analysis V8 (dbGaP Accession phs000424.v8.p2) with Median gene-level TPM by tissue using this link:
https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz

```{r libraries}
library(vroom)
library(dplyr)
library(stringr)
library(tidyr)
library(forcats)
library(ggplot2)
```

## Explore GTEx and make FIG 8
Import the data to R
```{r import median gene expression}
gtex_median_v8 <- vroom("GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct", skip = 2) %>% 
    filter(str_detect(Description, "PIWIL")) %>% 
    pivot_longer(cols = - c(Name, Description), names_to = "Tissue", values_to = "MEDIAN_TPM") %>% 
    mutate(Tissue = str_replace_all(Tissue, "-", "_") %>% str_replace_all(" ", "")) %>%
    filter(!str_detect(Tissue, "Cells"))

mean_med <- gtex_median_v8 %>% 
    separate(col = Tissue, into = c("Tissue", "subTissue"), sep = "_") %>% 
    group_by(Tissue, Description) %>% 
    summarise(medianTPM = median(MEDIAN_TPM)) %>% 
    ungroup() %>% 
    arrange(desc(medianTPM)) %>% 
    mutate(Tissue = as_factor(Tissue))
```

Make the figure 8
```{r}
tiff("FIG_8_GTEx_PIWIL_median_tpm.tiff", width = 12, height = 8, units = 'in', res = 300)
mean_med %>% 
    ggplot() +
    geom_col(mapping = aes(x = Description  , y = medianTPM, fill = Description))+
    facet_wrap(facets = "Tissue") +
    theme_bw(base_size = 12) + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    xlab("Gene") +
    ylab("Median TPM expression") +
    ggtitle("Bulk Tissue gene expression for PIWIL genes")
dev.off()
```

session info and restart R
```{r}
sessionInfo()
```

>R version 4.1.2 (2021-11-01)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19043)
Matrix products: default
locale:
[1] LC_COLLATE=English_United Kingdom.1252 
[2] LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252
[4] LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    
attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     
other attached packages:
[1] ggplot2_3.3.5 forcats_0.5.1 tidyr_1.2.0   stringr_1.4.0 dplyr_1.0.8  
[6] vroom_1.5.7  

## Explore recount TCGA and make FIG 10

# libraries
```{r warning=FALSE, include=FALSE}
library(recount3)
library(tidyverse)
library(plyranges)
library(DESeq2)
library(furrr)
library(tidybulk)
library(tidySummarizedExperiment)
library(ggrepel)
library(GGally)
library(ggpubr)
```

# get TCGA data
```{r}
human_projects <- recount3::available_projects()
project_info <- filter(human_projects, str_detect(file_source,"tcga")) 
projects <- project_info %>% 
    pull(project) %>% 
    set_names()
```

# pick one dataset and make rse object
```{r}
#rse_gene_ACC <- project_info %>% 
#    filter(project == "ACC") %>% 
#    create_rse(type = "gene", jxn_format = "UNIQUE")

tissues <- map(projects, ~ project_info %>% filter(project == .x) %>% 
                   create_rse(type = "gene",
                   jxn_format = "UNIQUE"))
```

# scale and transform counts to TPM
```{r}
#assayNames(rse_gene_ACC)
#assay(rse_gene_ACC, "counts") <- transform_counts(rse_gene_ACC)
#assays(rse_gene_ACC)$TPM <- recount::getTPM(rse_gene_ACC)
#colSums(assay(rse_gene_ACC, "TPM")) / 1e6 ## Should all be equal to 1


for(i in seq_along(tissues)){
assay(tissues[[i]], "counts") <- transform_counts(tissues[[i]])
assays(tissues[[i]])$TPM <- recount::getTPM(tissues[[i]])
}

rm(human_projects)
gc()
tissues %>% write_rds("tissues_TCGA_list_rse.rds")
#tissues <- read_rds("projects/tissues_TCGA_list_rse.rds")
```

# pick PIWIL genes names 
```{r}
ensg_names_piwil <- tissues[[1]] %>% 
    rowRanges() %>% 
    filter(str_detect(gene_name,"PIWIL")) %>% 
    as_tibble() %>% 
    select(gene_id, gene_name)
```

# pick piwil TPM expression
```{r}
# piwil_exp <- assays(rse_gene_ACC)$TPM %>% 
#    as_tibble(rownames = "gene_id") %>% 
#    inner_join(ensg_names_piwil) %>% 
#    select(gene_id, gene_name, everything()) %>% 
#    pivot_longer(cols = -c(gene_id, gene_name), names_to = "external_id", values_to = "TPM")

#piwil_exp %>% group_by(gene_id, gene_name )%>% summarise(median_TPM=median(TPM)) 

piwil_exp <- lapply(tissues, function(x){assays(x)$TPM %>% 
        as_tibble(rownames = "gene_id") %>% 
        filter(gene_id %in% ensg_names_piwil$gene_id)}) %>% 
    bind_rows(.id = "project") %>% 
    inner_join(ensg_names_piwil, by = c("gene_id")) %>% 
    select(gene_id, gene_name, everything()) 

piwil_exp <- piwil_exp %>% 
    pivot_longer(cols = -c(gene_id, gene_name, project), 
                 names_to = "external_id", values_to = "TPM") %>% 
    drop_na() 

```

# coldata
```{r}
# metad <- colData(rse_gene_ACC) %>% 
#    as_tibble() %>% 
#   select(external_id, study, primary_site = tcga.gdc_cases.project.primary_site,
#           condition = tcga.cgc_sample_sample_type)

metad <- map(tissues, ~colData(.x) %>% 
    as_tibble() %>% 
    select(external_id, project = study, project_name = tcga.gdc_cases.project.name, 
           primary_site = tcga.gdc_cases.project.primary_site,
           condition = tcga.cgc_sample_sample_type,
           platform = tcga.gdc_platform, center = tcga.gdc_center.short_name)) %>% 
    bind_rows()

piwil_exp <- piwil_exp %>% left_join(metad)

piwil_exp %>% vroom::vroom_write("TCGA_PIWIL_expression.txt")
 
#piwil_exp <- vroom::vroom("projects/TCGA_PIWIL_expression.txt")
```

# keep datasets that have: "Primary Tumor" and "Solid Tissue Normal"
```{r}
filtered_projects <- piwil_exp %>% 
    filter(condition %in% c("Primary Tumor", "Solid Tissue Normal")) %>% 
    count(project, condition) %>% 
    count(project) %>% 
    filter(n > 1) %>% 
    pull(project)

filtered_projects # 24 projects

piwil_exp %>% 
    filter(condition %in% c("Primary Tumor", "Solid Tissue Normal"),
           project %in% filtered_projects) %>% 
    count(platform) # 2 platforms

piwil_exp %>% 
    filter(condition %in% c("Primary Tumor", "Solid Tissue Normal"),
           project %in% filtered_projects) %>% 
    count(center) # 3 centers

piwil_exp <- piwil_exp %>% 
    filter(condition %in% c("Primary Tumor", "Solid Tissue Normal"),
           project %in% filtered_projects)

tissues <- tissues[filtered_projects]
```

# keep only the columns for DIF EXPR
```{r}
plan(multisession, workers = 2)

tissues <- future_map(tissues, ~.x %>% 
                        tidySummarizedExperiment::select(.feature:study,
                             tcga.gdc_cases.project.name,
                             tcga.gdc_cases.project.primary_site,
                             tcga.cgc_sample_sample_type,
                             tcga.gdc_platform, 
                             tcga.gdc_center.short_name) %>% 
                        tidySummarizedExperiment::select(-raw_counts) %>% 
                        tidySummarizedExperiment::rename(
                             project_name = tcga.gdc_cases.project.name,
                             primary_site = tcga.gdc_cases.project.primary_site,
                             condition = tcga.cgc_sample_sample_type,
                             platform = tcga.gdc_platform, 
                             center = tcga.gdc_center.short_name) %>% 
                        tidySummarizedExperiment::mutate(
                             condition = str_replace_all(condition, " ", "_")) %>% 
                        left_join(ensg_names_piwil, by = c(".feature" = "gene_id")))

# remove non standard samples
tissues <- future_map(tissues, ~.x %>%
                          tidySummarizedExperiment::filter(condition %in% c("Primary_Tumor", 
"Solid_Tissue_Normal")))
## save the list
tissues %>% write_rds("projects/tissues_TCGA_list_rse_for_DE.rds")

## tissues  <- read_rds("projects/tissues_TCGA_list_rse_for_DE.rds")
## Specify design and use tidybulk 
plan(multisession, workers = 2)
explore_batch <- tissues %>%  
    future_map(~.x %>% count(condition, project_name, primary_site, platform, center)) %>% 
    bind_rows(.id = "study") 

# edger
edger_get_DE <- function(x){
   
    
    dgList <- edgeR::SE2DGEList(x)

    group <- factor(dgList$samples$condition)
    keep.exprs <- filterByExpr(dgList, group = group)
    dgList <- dgList[keep.exprs,]
    dgList <- calcNormFactors(dgList)

    design <- model.matrix(~0 + dgList$samples$condition)
    colnames(design) <- colnames(design) %>% 
    str_remove("dgList\\$samples\\$condition") 
    rownames(design) <- rownames(dgList$samples)
    
    con_mat <- makeContrasts( Tumour_Normal = Primary_Tumor - Solid_Tissue_Normal,
                              levels = design)
    
    dgList <- estimateDisp(dgList, design, robust=TRUE)
    dgList <- glmQLFit(dgList, design, robust=TRUE)
    dgList <- glmQLFTest(dgList, contrast = con_mat[,"Tumour_Normal"])
    
    return(dgList)
}

## model matrix with batch
edger_get_DE_batch <- function(x){
   
    
    dgList <- edgeR::SE2DGEList(x)

    group <- factor(dgList$samples$condition)
    keep.exprs <- filterByExpr(dgList, group = group)
    dgList <- dgList[keep.exprs,]
    dgList <- calcNormFactors(dgList)

    design <- model.matrix(~0 + dgList$samples$condition + dgList$samples$platform)
    colnames(design) <- colnames(design) %>% 
    str_remove("dgList\\$samples\\$condition") %>% 
    str_remove("dgList\\$samples\\$platform")
    rownames(design) <- rownames(dgList$samples)
    
    con_mat <- makeContrasts( Tumour_Normal = Primary_Tumor - Solid_Tissue_Normal,
                              levels = design)
    
    dgList <- estimateDisp(dgList, design, robust=TRUE)
    dgList <- glmQLFit(dgList, design, robust=TRUE)
    dgList <- glmQLFTest(dgList, contrast = con_mat[,"Tumour_Normal"])
    
    return(dgList)
}
```

## BLCA, BRCA, CESC, CHOL, ESCA, GBM, HNSC, KICH, KIRC, KIRP, LIHC, LUAD, LUSC, PAAD, PCPG, PRAD, SARC, THCA, THYM
```{r}
no_batch_tissues <- c("BLCA", "BRCA", "CESC", "CHOL", "ESCA", "GBM", "HNSC", 
                      "KICH", "KIRC", "KIRP", "LIHC", "LUAD", "LUSC", "PAAD", 
                      "PCPG", "PRAD", "SARC", "THCA", "SKCM", "THYM")
tictoc::tic()
no_batch_tissues_DE <- tissues[no_batch_tissues] %>% 
    future_map(~edger_get_DE(.x))
tictoc::toc()
```

# COAD, READ, STAD, UCEC
```{r}
batch_tissues <- c("COAD", "READ", "STAD", "UCEC")
tictoc::tic()
batch_tissues_DE <- tissues[batch_tissues] %>% 
    future_map(~mutate(.x, platform = str_replace(platform, " ", "_")) %>% edger_get_DE_batch())
tictoc::toc()
```

# join the datasets
```{r}
tissues <- c(batch_tissues_DE, no_batch_tissues_DE) 
tissues %>% write_rds("projects/tissues_TCGA_dge_list_edger_DE.rds") 
# tissues <- read_rds("projects/tissues_TCGA_dge_list_edger_DE.rds") 
rm(batch_tissues_DE, no_batch_tissues_DE)
gc()
plan(multisession, workers = 1)
```

# check DE per tissue
```{r}
tissues_tidy <- tissues %>% 
    map(~.x %>% topTags(object = ., n = nrow(.), 
           adjust.method = "BH", sort.by = "PValue", p.value = 1) %>%
    .$table %>% 
    as_tibble(rownames = "smallRNA") %>% filter(!is.na(gene_name)) %>% 
    select(gene_name:FDR)) %>% 
    bind_rows(.id = "study")
```

# get log cpm values for plotting
```{r}
lcpm_piwil <- tissues %>% 
    map(~cpm(.x, log = TRUE, prior.count = 4) %>% 
            as_tibble(rownames = "ENSBLID") %>% 
            filter(ENSBLID %in% ensg_names_piwil$gene_id) %>% 
            pivot_longer(cols = -ENSBLID, names_to = "external_id", values_to = "lcpm") %>% 
            left_join(as_tibble(.x$samples))) %>% 
    bind_rows() %>% 
    left_join(ensg_names_piwil, by = c("ENSBLID" = "gene_id"))
```

# custom theme for the plots
```{r}
# Use colourblind-friendly colours
friendly_cols <- dittoSeq::dittoColors() %>% sample(100, replace = FALSE)

# Set theme
custom_theme <-
  list(
    scale_fill_manual(values = friendly_cols),
    scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 12),
        legend.position = "none",
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1)
      )
  )
```

# pick studies with more than 2 normal samples
```{r}
sel_studies <- lcpm_piwil %>% 
    filter(condition == "Solid_Tissue_Normal") %>% 
    count(study, condition, gene_name) %>% 
    filter(n > 2) %>% 
    count(study) %>% 
    pull(study)
```

# make violin plot
```{r}

my_comparisons <- list(c("Primary_Tumor", "Solid_Tissue_Normal"))
# PIWIL1

p <- lcpm_piwil %>%  
    filter(study %in% sel_studies, gene_name == "PIWIL4") %>% 
    mutate(study = as.factor(study), condition = as_factor(condition)) %>%
    ggviolin(x = "condition", y = "lcpm", fill = "study",
             add = "boxplot", 
             add.params = list(fill = "white")) +
    facet_wrap(~study, nrow = 3) + 
    stat_compare_means(comparisons = my_comparisons, label = "p.signif")+
    stat_compare_means(label.y = 10) +
    custom_theme +# Add significance levels
    ggtitle('PIWIL4')+
    theme(plot.title = element_text(hjust = 0.5)) 

pdf( file = "projects/PIWIL4_comparison_TCGA_lcpm.pdf", title = "PIWIL4", 
     paper = "a4r", width = 12, height = 9)
p
dev.off()

```
-   This project contains the scripting part of the Doctoral Dissertation of **Konstantinos Geles** with doi:
